bftengine/src/bcstatetransfer/Messages.hpp

```diff
@@ -40,8 +40,12 @@ class MsgType {
 };
 
 struct BCStateTranBaseMsg {
-  BCStateTranBaseMsg(uint16_t type) : type(type) {}
+  BCStateTranBaseMsg(uint16_t type) : BCStateTranBaseMsg(type, false) {}
+  BCStateTranBaseMsg(uint16_t type, bool isIncomingMsg) : type(type), is_incoming_msg_(isIncomingMsg) {}
   uint16_t type;
+  // this flag is for monitoring messages buffer allocs and releases so it's mutable eventhough incoming ST msgs objs
+  // are const during handling
+  mutable bool is_incoming_msg_ = false;
```

Comment:
to answer all your points:

1. will do
2. will init in ctor
3. in the state transfer msg handler we retrieve a single msg as const
4. since ST msgs are encapsulated in MessageBase and sent to other replicas, it is passed on the replica network. only if different replicas are compiled differently it can cause a problem. It's easy to change to uint_8 so I think it's best.